src/conn_state.c

```diff
@@ -0,0 +1,163 @@
+/*
+ * Copyright 2019 Xaptum, Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#include "conn_state.h"
+
+#include <errno.h>
+#include <fcntl.h>
+#include <string.h>
+#include <unistd.h>
+
+#include <arpa/inet.h>
+#include <net/if.h>
+#include <netinet/in.h>
+#include <sys/socket.h>
+
+#include "context.h"
+#include "log.h"
+
+static
+int
+connect_udp_socket(char* udp_address, int udp_length, struct addrinfo* connect_addr)
```

Comment:
This function does too much.  It connects a UDP socket (per its name), switches on the address family, gets the local address of the socket, converts it to textual form, and returns that.

I suggest splitting this into multiple functions:

- [ ] abstract the UDP socket in the same way we did the tcp socket in `tcp.h`:
```
# udp.h

struct enftun_udp
{
   int fd; // file descriptor for the underlying udp socket
   
   struct sockaddr local_addr;
   struct sockaddr remote_addr;
}

enftun_udp_connect_addr(struct enftun_udp* udp, struct addrinfo* addr); // also sets the local_addr and remote_addr params
enftun_udp_close(struct enftun_udp* udp);
```

- [ ] rather than converting the IP address from the native `struct sockaddr` to a textual form to compare, write a comparison function directly.  Much faster than generating the textual strings.

- [ ] update the `udp_check` function to use those new functions.  Maybe give it a more precise name.